### PR TITLE
fix(voting): improves voting behavior, logging, and reduces excessive vote sending

### PR DESCRIPTION
**Pull request by justinsnapp**
## For reviewers

This is a clean combining of PRs #956 and #959

## Version Notes ##
{ major | minor | patch }

**NOTE: Include the version notes in the squash commit message**

### CODE REVIEW

File: src/network/taraxa_capability.cpp [hunk 1]
```diff
@@ -592,31 +607,72 @@ void TaraxaCapability::interpretCapabilityPacketImpl(NodeID const &_nodeID, unsi
...
       for (size_t i = 0; i < next_votes_count; i++) {
         Vote next_vote(_r[i].data().toBytes());
-        LOG(log_nf_next_votes_sync_) << "Received PBFT next vote " << next_vote.getHash();
+        if (next_vote.getRound() != peer_pbft_round - 1) {
+          LOG(log_er_next_votes_sync_) << "Received next votes bundle with unmatched rounds from " << _nodeID
+                                       << ". The peer may be a malicous player, will be disconnected";
+          host->disconnect(_nodeID, p2p::UserReason);
+
+          break;
+        }
```
Comment (gaoqi8649) [inline src/network/taraxa_capability.cpp:619]: This will always true. because line 609

File: src/network/taraxa_capability.cpp [hunk 2]
```diff
@@ -592,31 +607,72 @@ void TaraxaCapability::interpretCapabilityPacketImpl(NodeID const &_nodeID, unsi
...
+        auto next_vote_hash = next_vote.getHash();
+        LOG(log_nf_next_votes_sync_) << "Received PBFT next vote " << next_vote_hash;
 
         next_votes.emplace_back(next_vote);
```
Comment (gaoqi8649) [inline src/network/taraxa_capability.cpp:623]: I think should mark peer as known here.
Don't mark peer known votes at "if (pbft_current_round < peer_pbft_round)" situation

File: src/network/taraxa_capability.cpp [hunk 3]
```diff
@@ -592,31 +607,72 @@ void TaraxaCapability::interpretCapabilityPacketImpl(NodeID const &_nodeID, unsi
...
+        }
       } else if (pbft_current_round == peer_pbft_round) {
         // Update previous round next votes
         auto pbft_2t_plus_1 = db_->getPbft2TPlus1(pbft_current_round - 1);
         if (pbft_2t_plus_1) {
+          // Update our previous round next vote bundles...
           next_votes_mgr_->updateWithSyncedVotes(next_votes, pbft_2t_plus_1);
+          for (auto const &v : next_votes) {
+            peer->markVoteAsKnown(v.getHash());
+          }
```
Comment (gaoqi8649) [inline src/network/taraxa_capability.cpp:651]: Then should remove this

File: src/network/taraxa_capability.cpp [hunk 4]
```diff
@@ -1742,39 +1808,50 @@ void TaraxaCapability::requestPbftNextVotes(NodeID const &peerID, uint64_t const
...
+  for (auto const &next_vote : send_next_votes_bundle) {
     s.appendRaw(next_vote.rlp());
-    LOG(log_nf_next_votes_sync_) << "Send out next vote " << next_vote.getHash() << " to peer " << peerID;
+    LOG(log_dg_next_votes_sync_) << "Send out next vote " << next_vote.getHash() << " to peer " << peerID;
+  }
+
+  if (sealAndSend(peerID, PbftNextVotesPacket, move(s))) {
+    LOG(log_nf_next_votes_sync_) << "Send out size of " << send_next_votes_bundle.size() << " PBFT next votes to "
+                                 << peerID;
+    auto peer = getPeer(peerID);
```
Comment (gaoqi8649) [inline src/network/taraxa_capability.cpp:1825]: check peer is not nullptr

File: src/network/taraxa_capability.cpp (hunk 1 above)
Comment (justinsnapp) [reply src/network/taraxa_capability.cpp:619]: I disagree @gaoqi8649 ,  609 only checks the first vote received relative to node's round.   This line checks that all votes received are in the same round.

File: src/network/taraxa_capability.cpp (hunk 2 above)
Comment (justinsnapp) [reply src/network/taraxa_capability.cpp:623]: Fixed in #969

File: src/network/taraxa_capability.cpp (hunk 3 above)
Comment (justinsnapp) [reply src/network/taraxa_capability.cpp:651]: Fixed in #969

File: src/network/taraxa_capability.cpp (hunk 4 above)
Comment (justinsnapp) [reply src/network/taraxa_capability.cpp:1825]: Fixed in #969

**Pull request merged**